src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+  on `s` over `f`. -/
+lemma finsum_in_insert_of_eq_zero_if_not_mem (h : a ∉ s → f a = 0) :
+  ∑ᶠ i in (insert a s), f i = ∑ᶠ i in s, f i :=
+begin
+  by_cases hs : ((s ∩ function.support f).finite),
+    { by_cases hm : a ∈ s,
+        { simp_rw set.insert_eq_of_mem hm },
+        { rw [finsum_in_insert' f hm hs, h hm, zero_add] } },
+    { rw [finsum_in_eq_zero_of_infinite hs, finsum_in_eq_zero_of_infinite],
+      intro hs', apply hs,
+      by_cases hm : a ∈ function.support f,
+        { apply set.finite.subset hs',
+          finish },
+        { rwa set.insert_inter_of_not_mem hm at hs' } }
+end
+
+/-- If `f a = 0`, then the sum on `insert a s` over the function `f` equals the sum on `s` over `f`.
+  -/
+@[simp] lemma finsum_insert_zero (h : f a = 0) :
+  ∑ᶠ i in (insert a s), f i = ∑ᶠ i in s, f i :=
+finsum_in_insert_of_eq_zero_if_not_mem (λ _, h)
+
+/-- The sum on a singleton `{a}` over the function `f` is `f a`. -/
+@[simp] lemma finsum_singleton : ∑ᶠ i in {a}, f i = f a :=
+begin
+  rw [finsum_in_eq_finset_sum''' f (set.finite_singleton a), set.finite.to_finset,
+      ← @finset.sum_singleton _ _ a f _],
+  congr, ext, simp
+end
+
+/-- The sum on the set of two unique elements `{a, b}` over the function `f` equals `f a + f b`. -/
+@[simp] lemma finsum_pair (h : a ≠ b) : finsum_in {a, b} f = f a + f b :=
+begin
+  rw [finsum_in_eq_finset_sum''' f (show set.finite {a, b}, by simp),
+      set.finite.to_finset, ← finset.sum_pair h],
+  congr, ext, simp
+end
+
+/-- The sum on the image `g '' s` over the function `f` equals the sum on `s` over `f ∘ g` given `g`
+  is injective on `s`. -/
+@[simp] lemma finsum_in_image' {s : set β} {g : β → α} (hs : s.finite) (hg : set.inj_on g s) :
+  ∑ᶠ j in (g '' s), f j = ∑ᶠ i in s, (f ∘ g) i :=
+begin
+  rw [finsum_in_eq_finset_sum''' (f ∘ g) hs, finsum_in_eq_finset_sum''' f (set.finite.image g hs),
+    ←finset.sum_image
+    (λ x hx y hy hxy, hg (set.finite.mem_to_finset.1 hx) (set.finite.mem_to_finset.1 hy) hxy)],
+  congr, ext, simp
+end
+
+/-- A more general version of `finsum_in_image'` that does not require `s` to be finite. -/
+@[simp] lemma finsum_in_image {s : set β} {g : β → α}
+  (hg : set.inj_on g s) : finsum_in (g '' s) f = finsum_in s (f ∘ g) :=
+begin
+  by_cases hs : (g '' s ∩ function.support f).finite,
+    { rw [finsum_in_eq_finset_sum _ hs, finsum_in_eq_finset_sum _
+            ((set.image_inter_support_finite_iff hg).1 hs),
+        ← finsum_in_eq_finset_sum''', ← finsum_in_eq_finset_sum''',
+        ← (@finsum_in_image' _ _ _ _ f (s ∩ function.support (f ∘ g)) g
+            ((set.image_inter_support_finite_iff hg).1 hs)
+          (set.inj_on.mono (set.inter_subset_left s _) hg))],
+        { congr, ext y, refine ⟨_, by finish⟩,
+          rintro ⟨⟨x, hx, rfl⟩, hy⟩,
+          exact ⟨x, ⟨hx, hy⟩, rfl⟩ } },
+    { rw [finsum_in_eq_zero_of_infinite hs, finsum_in_eq_zero_of_infinite],
+      intro hs', apply hs,
+      rwa set.image_inter_support_finite_iff hg }
+end .
```

Comment:
```suggestion
begin
  by_cases hs : (g '' s ∩ function.support f).finite,
  { rw [finsum_in_eq_finset_sum _ hs, finsum_in_eq_finset_sum _
          ((set.image_inter_support_finite_iff hg).1 hs),
      ← finsum_in_eq_finset_sum''', ← finsum_in_eq_finset_sum''',
      ← (@finsum_in_image' _ _ _ _ f (s ∩ function.support (f ∘ g)) g
          ((set.image_inter_support_finite_iff hg).1 hs)
        (set.inj_on.mono (set.inter_subset_left s _) hg))],
    { congr, ext y, refine ⟨_, by finish⟩,
      rintro ⟨⟨x, hx, rfl⟩, hy⟩,
      exact ⟨x, ⟨hx, hy⟩, rfl⟩ } },
  { rw [finsum_in_eq_zero_of_infinite hs, finsum_in_eq_zero_of_infinite],
    intro hs', apply hs,
    rwa set.image_inter_support_finite_iff hg }
end .
```

Reply:
We rewrote the proof incorporating the previous lemma.